pkg/platform/runtime/buildscript/buildscript.go

```diff
@@ -119,9 +140,35 @@ func (s *Script) String() string {
 	case s.Expr.Let.In.Name != nil:
 		buf.WriteString(*s.Expr.Let.In.Name)
 	}
+
 	return buf.String()
 }
 
+// prepareForOutput prepares this script's buildexpression for output in buildscript format.
+// For example, substitute the solve node's "at_time" timestamp with a variable reference.
+// Since this function directly modifies the Expr field, callers of this function must balance it
+// with a call to `prepareForUse()` to restore the Expr to a useful state.
+func (s *Script) prepareForOutput() {
+	if s.atTime != nil {
+		_, err := s.Expr.SetDefaultTimestamp()
+		if err != nil {
+			multilog.Error("Failed to set default timestamp: %v", err)
+		}
+	}
+}
+
+// prepareForUse prepares this script's buildexpression for use in a buildplanner context.
+// For example, the solve node's "at_time" will not contain a variable reference if this script
+// has a top-level assignment for it.
+func (s *Script) prepareForUse() {
+	if s.atTime != nil {
+		err := s.Expr.MaybeUpdateTimestamp(*s.atTime)
+		if err != nil {
+			multilog.Error("Failed to possibly update at time: %v", err)
```

Comment:
We should return errors here. It's ok that this breaks things for the user, that's our signal that something needs fixing.